add flatten parameter module

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66578

flatten parameters for performance optimization and handle the case when grad ready order is different or there are unused parameters among ranks. when there is no param to be sharded in the FSDP instance (usually root), the flatten wrapper module's flat_param is None.

Differential Revision: [D31625194](https://our.internmc.facebook.com/intern/diff/D31625194/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

zhaojuanmao committed:
add flatten parameter module

flatten parameters for performance optimization and handle the case when grad ready order is different or there are unused parameters among ranks. when there is no param to be sharded in the FSDP instance (usually root), the flatten wrapper module's flat_param is None.

Differential Revision: [D31625194](https://our.internmc.facebook.com/intern/diff/D31625194/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add flatten parameter module"


flatten parameters for performance optimization and handle the case when grad ready order is different or there are unused parameters among ranks. when there is no param to be sharded in the FSDP instance (usually root), the flatten wrapper module's flat_param is None.

Differential Revision: [D31625194](https://our.internmc.facebook.com/intern/diff/D31625194/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add flatten parameter module"


flatten parameters for performance optimization and handle the case when grad ready order is different or there are unused parameters among ranks. when there is no param to be sharded in the FSDP instance (usually root), the flatten wrapper module's flat_param is None.

Differential Revision: [D31625194](https://our.internmc.facebook.com/intern/diff/D31625194/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add flatten parameter module"


flatten parameters for performance optimization and handle the case when grad ready order is different or there are unused parameters among ranks. when there is no param to be sharded in the FSDP instance (usually root), the flatten wrapper module's flat_param is None.

Differential Revision: [D31625194](https://our.internmc.facebook.com/intern/diff/D31625194/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

mrshenli reviewed (commented):
Lint failure is real:

```
./torch/distributed/_fsdp/flatten_params_wrapper.py:178:139: B950 line too long (138 > 120 characters)
```

mrshenli reviewed (approved):
Stamp to unblock. It will be better if we can add back the tests for flatten parameters

zhaojuanmao committed:
Update on "add flatten parameter module"


flatten parameters for performance optimization and handle the case when grad ready order is different or there are unused parameters among ranks. when there is no param to be sharded in the FSDP instance (usually root), the flatten wrapper module's flat_param is None.

Differential Revision: [D31625194](https://our.internmc.facebook.com/intern/diff/D31625194/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]
